### PR TITLE
feat(lang): support str substring check with `in` / `not in` (#1032)

### LINKED ISSUE #1032: improvement(lang): `in` operator should support substring check for `str`
## Summary

The `in` operator works for `List`, `Map`, and `Set` membership tests, but raises a compile-time error when the right-hand side is a `str`. Users must call `contains()` instead, which is inconsistent with the syntax used for collections.

## Reproduction

```ry
print(3 in [1, 2, 3])            # true  — works
print("a" in {"a": 1, "b": 2})  # true  — works (key lookup)
print(20 in {10, 20, 30})        # true  — works

print("world" in "hello world")
# error: 'in' operator requires a set, list, or map on the right side
```

## Expected Behaviour

```ry
print("world" in "hello world")   # true
print("xyz"   in "hello world")   # false
print("xyz" not in "hello world") # true
```

This would be consistent with Python and would allow `str` to be used uniformly wherever membership tests appear (e.g. `case` guards, `filter` predicates).

## Proposed Implementation

In the type checker / codegen, when the right operand of `in` / `not in` has type `str`, lower it to a `contains(rhs, lhs)` call.

### PR DESCRIPTION
**Pull request by t0k0sh1**
## Summary

- `"world" in "hello world"` now evaluates to `true`; `"xyz" in "hello world"` to `false`
- `not in` is the negation: `"xyz" not in "hello world"` → `true`
- Empty needle `"" in s` is always `true` (matches Python / existing `contains` semantics)
- `any`-typed LHS holding a `str` is transparently unwrapped via `unwrapFromAny`
- Non-`str` LHS with a `str` RHS is a compile error

## Implementation

Adds a new `str` branch in `emitExprVariant` (`src/codegen_expr.cpp`) after the existing Set/Map/List branches, lowering to the existing `__ry_str_find_byte` runtime function used by `contains()`.

Dispatch order: user overload (`tryOperatorCall`) → Set → Map → List → **str** → error.

`isStringValue(container)` correctly excludes collection/resource-metadata pointers, so the str branch does not interfere with existing container branches.

No parser, lexer, AST, `share/std`, or runtime changes needed.

## Test plan

- [x] `CodeGenTest.InNotInOperatorsStr` — 9 C++ cases: true/false, `not in`, empty needle, full match, non-str rejection, `any` LHS widening
- [x] `tests/spec/strings.test.ry` — 2 new `it` blocks: `in` operator (5 cases) and `not in` operator (2 cases)
- [x] `./build/ry_tests` — 1773 tests pass
- [x] `./build/ry test -p` — spec tests pass (serial mode; parallel runner has pre-existing flakiness tracked in #1088)
- [x] ASan+UBSan: 1773 C++ tests + 131 Ry spec files clean
- [x] TSan: 1773 C++ tests clean

Closes #1032

## Summary by CodeRabbit

* **New Features**
  * `in` and `not in` now perform substring membership checks for strings (e.g., `"world" in "hello world"` → true); empty-string needle always matches.

* **Documentation**
  * Added KNOWLEDGE.md and updated operator reference and changelog to describe string membership semantics and edge cases.

* **Tests**
  * Added unit and spec tests covering substring membership, empty-needle behavior, type-error cases, and `any`-typed needle handling.

### CODE REVIEW

File: src/codegen_expr.cpp [hunk 1]
```diff
@@ -1582,7 +1582,28 @@ llvm::Value *CodeGen::emitExprVariant(const std::unique_ptr<BinaryExpr> &e) {
...
+        // Try str (substring check) — #1032
+        if (isStringValue(container)) {
+            if (!isStringValue(elem)) {
+                if (isAnyType(elem->getType()) && canAnyHoldType(ptrTy_))
+                    elem = unwrapFromAny(elem, ptrTy_);
+                else
+                    codegenError("'" + e->op + "' operator: left side must be str when right side is str");
+            }
+            llvm::Value *hlen = emitStringByteLen(container);
+            llvm::Value *nlen = emitStringByteLen(elem);
```
[the remark this answers is not in the document]
Comment (t0k0sh1) [reply src/codegen_expr.cpp:1598]: Thanks for the thorough analysis. Let me address both dimensions of safety here.

**Type-system invariant (compile-time layer)**: In the Ry type system `any` can only hold `int`, `float`, `bool`, or `str` — enforced by `wrapInAny()` in `codegen_any.cpp`, which calls `codegenError` if `isNonStrPointer(val)` is true (i.e. the value carries List/Map/Set metadata). So `x: any = [1]` is already a compile error before reaching the `in` operator. A new regression test `InStrAnyListRejection` confirms this invariant.

**Runtime safety net (second layer)**: `unwrapFromAny(elem, ptrTy_)` is not a blind cast — it emits LLVM IR that extracts the tag field from the `{i64, [8 x i8]}` struct and checks `tag == RyAnyTag::Str` (see `codegen_any.cpp:110–126`). On mismatch it calls `emitRuntimeError("runtime error: any type mismatch (expected str)\n")` and is marked `unreachable`. A List/Map/Set pointer cannot slip through to `emitStringByteLen` even in theory.

**Change made**: Added a 4-line safety comment in `src/codegen_expr.cpp` at the `unwrapFromAny` call site explaining both layers, and added the `InStrAnyListRejection` compile-error regression test to `tests/test_codegen_collection.cpp`.
